### ISSUE TITLE
Suggestion: vim mode

### ISSUE THREAD

**Issue by pierd** [Canop/broot#185]
I was thinking about a mode that would be more familiar to vim users. By default the keys would be commands:
- `j`/`k` or down/up arrow - moving down/up
- enter or `i` - (on a file) open it
- enter - (on a directory) enter it exiting broot (aka `cd` in parent shell)
- `i` - (on a directory) enter it staying in broot
- `u` - go back (undo last enter directory) OR go up directory tree
- `/` - enter search mode
- `h`/`l` - moving up/down the tree (similar to https://github.com/Canop/broot/issues/179)
- `:` - start typing a verb
... and possibly more.

Search mode would be pretty much like the current broot behaviour with the only change being:
- enter or esc - go back to navigation mode.

What do you think? I think that would be something I would use by default.

I'm happy to start implementing it and/or preparing a prototype for testing the UX. I envision having this mode accessible with a command line argument (`--vim`, `--vim-mode`) or a config file setting.

**Comment by Canop** [comment]
This is something that is often discussed. Thanks for setting it as a separate issue which could be the basis for a discussion for the complete specification later.

Please don't try developing it now: there are already several branchs in parallel in my code at the moment, some of them related to the input.

**Comment by Canop** [comment]
The esc key, today, serves many purposes, behind the Back umbrella, the most useful ones being removing the current search. Using this key for going to command mode is a problem and means we find an alternative key for Back.

**Comment by vanhtuan0409** [comment]
Backspace would be an option

**Comment by aenglisc** [comment]
Ctrl + [ is idiomatic, I'd go with that.

**Comment by pierd** [comment]
yip, `esc` == `^[`

**Comment by Bintux** [comment]
> yip, `esc` == `^[`

On a german keyboard this is a very cumbersome shortcut as you have to press  Ctrl +right Alt + 8

**Comment by gorakhargosh** [comment]
Vim mode would be way too cool!

**Comment by Canop** [comment]
I still don't think this mode would be a good idea for broot (and I say that as an avid vi user) but it's possible to try. The current version of broot makes it easier (for a courageous and competent Rust coder) to do such experimentation.

**Comment by AXGKl** [comment]
search only after `/` and vim style keyboard navigation (min. `j`, `k`, `gg`, `G`), that would give this fubared year at least *something* positive ;-)

**Comment by ddrcode** [comment]
Yet another vote for Vim mode. Please note that tools which offer slightly similar functionality (to some extent) - like nnn or ranger - come vim key bindings. I think broot is a fantastic tool and I'm impressed by its concept. The lack of vim modes is my only problem here. I use Planck keyboard, without arrow keys (in the main layer), which makes vim-like navigation my preferred option.

**Comment by Canop** [comment]
To me it looks like the goal is to be able to have no-modifier key shortcuts as we have in vim.

We can imagine there are two modes in broot:

1) input focused (mode I)
2) input not focused (mode C)

Right now, the only mode in broot is the first one, mode I (as "insertion"). We could say there's just no way yet to activate the second one (mode C, as "command").

A new toggle could enable the C mode and introduce two new verbs, dedicated to going from one mode to the other one.

It seems natural to bind the entry in "insertion" mode to keys like "i", "/", " ".

And going to mode C could be bind to 'esc'.

I think I could have a flag in config allowing this modality. It doesn't have to explicitly refer to vim.

All verbs would be optionally tied to one mode.

In some cases the shortcut would automatically be enabled or disabled: you can't have no-modifier normal letters key bindings in mode I.

A consequence is that the activation of the global modality flag would be enough to enable this without having to maintain two config files (of course we *can* have two config files).

 *Does that look clear ? Feedback welcome*

**Comment by ddrcode** [comment]
It looks clear and the mode switch verb should be sufficient to achieve modality - the rest should depend on configuration. I think there should be also an option to configure the initial mode: for most of users the insert mode at start should be just fine, bu some (vimers?) may prefer to start with command mode.

**Comment by Canop** [comment]
When the toggle is on, the initial mode would probably be the command mode. But I'll have to test and feel whether it "works".

**Comment by Canop** [comment]
The last commit introduces a "vim mode".

Explanations here: https://github.com/Canop/broot/blob/master/vim_mode.md

**I need feedback!**

**Comment by 20manas** [comment]
The bindings which have been added for input/insert mode are `/`, `:` and ` `. I would suggest the following default bindings for the vim mode:

1. `i` or `/`: enter input/insert mode.
2. `:`: enter input/insert mode and type ":" in the bottom field (which is similar to how it works in vim).
3. `space` for input/insert seems unnecessary; instead it can be used for some other commonly used verb like `open_preview`.

**Comment by Canop** [comment]
@20manas Did you test ? There are already different behaviors for the `/`, ` ` and `:` keys. When you hit `:`, the `:` is added to the input if there's not already a verb. The space, whose usage in broot is the same, works the same. The reason I didn't include the `i` (but you can bind it with the config file) is to have the possibility to use it other ways (@pierd suggested its use to open files).

**Comment by 20manas** [comment]
Sorry, I didn't test. Never mind what I had said. The current bindings are great.

**Comment by vanhtuan0409** [comment]
Hi @Canop , can we consider clearing the last input after pressing `/` or `:` ?
The current behavior is:
- Pressing `:`, enter some command, pressing `esc` to return to command mode (no command executed)
- Pressing `/`, the last input is shown. Expected the last input should be clear, ready for a fresh new search

If this is not desired, can we add a config option to toggle this behavior

**Comment by ddrcode** [comment]
Thanks for implementing the modal mode, @Canop. I think we should abandon the name "vim mode", as you suggested in your md file - modal mode sounds much more accurate.

I have switched completely to it and have been using it for last three days. It feels natural, as I spend most of my time in vim/tmux. I configured it for Colemak equivalent of HJKL -> HNEI and I prefer the modal version over the "original" one. 

There are some tiny problems/inconsistencies, though:
* Typical vimer habit - to press `ESC` when unsure in which mode you are is problematic here, as `ESC` pressed in insert mode behaves differently (switches to command mode) than pressed in command mode (`:back`). I got trapped frequently by it. I wanted to override the original behaviour of `ESC` with _do nothing when in command mode_, but couldn't figure out how. That may be actually a suggestion for a small generic change - something like `:no-action` verb.
* The re-search behaves bit strange - as pointed out by @vanhtuan0409. Search for a phrase by typing `/phrase` - it will work fine at first, but pressing `/` character again produces previous search string followed by slash (`phrase/`). It should either clear the search or at lest avoid `/` (I'd prefer the latter)
* switching to insert mode by pressing space actually produces space character, which impacts search results

I have reviewed your commit related to modal mode and and it all looks logical and clean. I think the modal mode is worth adding to the official features of broot - after solving the problems mentioned above.

**Comment by ddrcode** [comment]
Here are my further considerations and observations (continuation of my comment from above).
First - please see my config file I use (for readability and understanding I converted it to qwerty settings from my original  Colemak-specific keymapping, so it uses standard vim-like keybinding):

```toml
modal=true

[[verbs]]
execution = ":focus"
key = "l"

[[verbs]]
execution = ":parent"
key = "h"

[[verbs]]
execution = ":line_down"
key = "j"

[[verbs]]
execution = ":line_up"
key = "k"

[[verbs]]
execution = ":select_first"
key = "g"

[[verbs]]
execution = ":select_last"
key = "G"

[[verbs]]
execution = ":back"
key = "u"

[[verbs]]
execution = ":page_down"
key = "ctrl-f"

[[verbs]]
execution = ":page_up"
key = "ctrl-b"

[[verbs]]
execution = ":close_panel_ok"
key = "q"

[[verbs]]
execution = ":panel_right"
key = "ctrl-l"

[[verbs]]
execution = ":panel_left"
key = "ctrl-h"

[[verbs]]
execution = ":mode_input"
key = "i"

[[verbs]]
invocation = "edit"
key = "e"
execution = "$EDITOR {file}"
leave_broot = true
```

I hope it's handy and _good-to-go_ for most of vim users. I'm not yet very convinced about using _undo_ (`u`) shortcut for for `:back` and I'm tempted to replace it with `ctrl-o` (prev buffer in vim).

One non-standard thing in the configuration is `g` for `:select_first`. Of course in vim it would be `gg`, but in broot we don't have key sequences for shortcuts. It's a topic on its own and I think it shouldn't be considered as part of this ticket - however  worth exploring the idea in the future (@Canop I'm happy to create a more descriptive ticket for it if you are interested, as I have some practical ideas of how it could be used). 

`G` (I also tried `shift-g`) doesn't work for `:select_first` - seems broot ignores shortcuts which are capital-case. 

One thing I really like in my configuration is mapping `q` as `:close_panel_ok`, which closes current panel (if multiple open) or quits from broot (if only one panel is open) - pretty much like splits in vim. 

One problem I found is related to mapping `ESC`. I wanted to do something like this
```toml
[[verbs]]
execution = ":mode_command"
key = "ESC"
```
but it throws an error: `Bad configuration: reserved key: Esc`. And then - I just assume - even without error it wouldn't work as broot ignores all single-key shortcuts in the insert mode. I think would be good to fix that and allow to overwrite `ESC`.

Finally - there is yet another problem which would be worth fixing - when I type a command from command mode (i.e. I type `:select_first`), broot will switch to insert mode and remain in that mode after the execution of the command. It's related to the fact that `:` is considered as a switch to input mode, which is not precisely correct.

The modal mode - even with the described problems - is very usable now. Once the problems fixed it will be a great addition to the tool.

**Comment by Canop** [comment]
> And then - I just assume - even without error it wouldn't work as broot ignores all single-key shortcuts in the insert mode
Not really. The ignored keys are the "character" ones. But ESC is a special beast, with a lot of very contextual meanings.

I've noted a few things in your comments. I'll have to go deeper when I have enough time.

**Comment by lstwn** [comment]
> `G` (I also tried `shift-g`) doesn't work for `:select_first` - seems broot ignores shortcuts which are capital-case.

Can confirm this with latest version `1.2.9`.

Btw, modal mode is great! Although, I perfectly understand the concerns about how broot is designed to be, modal mode great for usage with vim :)

**Comment by Canop** [comment]
I'll keep the modal mode in broot.

So I close this issue, as done.

Other related problems, wishes, etc. should thus be directed to new issues or (usually better at first) to the chat.

And I'll create a new issue for the `shift-g` problem that I had overlooked.

**Comment by Canop** [comment]
@ddrcode  @lstwn 
I've opened this issue: https://github.com/Canop/broot/issues/363

**Comment by Canop** [comment]
@theowenyoung `:mode_input` is an internal which, by design, can't be called by typing it. The mode changes on typing some specific characters.

> You enter input mode by typing one of those letters: (space), :, or /. You leave it with the escape key. You may add other bindings to the :mode_input and :mode_command verbs.

It may be confusing (I didn't find the right terminology I think) so I suggest you read this page: https://dystroy.org/broot/modal/

**Comment by theowenyoung** [comment]
Thanks, it works when I put it to  configuration file.

**Comment by Canop** [comment]
I know "modal mode" and "mode" weren't really clear...
